library/bignum_mod_raw.c

```diff
@@ -94,4 +94,61 @@ int mbedtls_mpi_mod_raw_write( const mbedtls_mpi_uint *A,
     }
 }
 
+int mbedtls_mpi_set_montgomery_constant_unsafe( const mbedtls_mpi_uint **X,
+                                                const mbedtls_mpi_uint *A,
+                                                size_t limbs )
+                                                
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    mbedtls_mpi N;
+    mbedtls_mpi RR;
+
+    mbedtls_mpi_init( &N ); 
+    mbedtls_mpi_init( &RR );
+
+    if ( A == NULL )
+        goto cleanup;
+
+    if ( !mbedtls_mpi_grow( &N,  limbs ) )
+        memcpy( N.p, A, sizeof(mbedtls_mpi_uint) *  limbs );
+    else
+        goto cleanup;
+
+    MBEDTLS_MPI_CHK( mbedtls_mpi_lset( &RR, 1 ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_shift_l( &RR, N.n * 2 * biL ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_mod_mpi( &RR, &RR, &N ) );
+    MBEDTLS_MPI_CHK( mbedtls_mpi_grow( &RR, N.n ) );
+    *X = RR.p;
+    RR.p = NULL;
+    ret = 0;
+
+cleanup:
+    mbedtls_mpi_free(&N);
+    mbedtls_mpi_free(&RR);
+    return( ret );
+}
+
+int mbedtls_mpi_mod_raw_from_mont_rep( mbedtls_mpi_uint *X,
+                                       const mbedtls_mpi_mod_modulus *m )
+{
+    mbedtls_mpi_uint one = 1;
+    mbedtls_mpi T;
+    mbedtls_mpi_init( &T );
+    mbedtls_mpi_core_montmul( X, X, &one, 1, m->p, m->limbs,
+                              m->rep.mont.mm, T.p );
+    mbedtls_mpi_free( &T );
+    return( 0 );
+}
+
+int mbedtls_mpi_mod_raw_to_mont_rep( mbedtls_mpi_uint *X,
+                                     const mbedtls_mpi_mod_modulus *m )
+{
+    mbedtls_mpi T;
+    mbedtls_mpi_init( &T );
+    mbedtls_mpi_core_montmul( X, X, m->rep.mont.rr, 1, m->p, m->limbs,
```

Comment:
You are indeed correct that this `1` needs to be updated.

Since we brought this subject, the actual limb size needed to represent RR is not always the same always `m->limbs`  We can have an 8 limbs modulo and `RR == 0x1` which would need just **one** limb. 

The implicit assumption is that `limbs_RR <= limbs_N`. It shouldn't affect our operations, but it may be worth to look into for optimisation?

Reply:
In theory yes, however unless the modulus has some peculiar structure, I would expect RR to need limbs_N in most cases.